File: 프로그래머스/lv2/131127. 할인 행사/할인 행사.py
from collections import defaultdict

def solution(want, number, discount):
    answer = 0
    cri = dict()
    
    for i, eachWant in enumerate(want):
        cri[eachWant] = number[i]
        
    for i in range(len(discount) - 9):
        tmpDict = defaultdict(int)

        for j in range(i, i + 10):
            tmpDict[discount[j]] += 1
    
        if tmpDict == cri:
            answer += 1
    
    return answer

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        int n = discount.length;
        
        Map<String, Integer> map = new HashMap<>();
        
        for (int i = 0; i < n; i++) {
            String dis = discount[i];
            
            if (i < 10) {
                map.put(dis, map.getOrDefault(dis, 0) + 1);
                if (i != 9) continue;
            }
            
            if (i > 9) {
                map.put(dis, map.getOrDefault(dis, 0) + 1);
                map.put(discount[i - 10], map.get(discount[i - 10]) - 1);
            }
            
            boolean flag = true;
            for (int j = 0; j < want.length; j++) {
                if (map.getOrDefault(want[j], 0) < number[j]) {
                    flag = false;
                    break;
                }
            }
            
            if (flag) answer += 1;
            
        }        
        
        return answer;
    }
}